### ISSUE TITLE
How to remove Field from redux form v6

### ISSUE THREAD

**Issue by swp44744** [erikras/redux-form#2325]
I am using redux form 6. I am creating dynamic Fields based on ajax response. if ajax response is changed new Fields gets created. and I dont know how to remove earlier Fields from redux form.

any suggestions.

**Comment by gustavohenke** [comment]
Use `unregisterField` action; from [the docs](http://redux-form.com/6.3.1/docs/api/ActionCreators.md/):

> #### `unregisterField(form:String, name:String)`
> Unregisters a field with the form.

Let me know whether this solves your problem.

**Comment by swp44744** [comment]
@gustavohenke thanks for syntax. I tried, it removed it from register fields but fileds & values section still have an entry.

**Comment by swp44744** [comment]
I tried different action creators that redux form support currently but none of them I could use to remove unregistered field from value and the field section.

**Comment by JakeLin1024** [comment]
how about re-initializing your form with 'new values' after you unregister fields...

It might work but not I don't think it's an ideal way.

**Comment by swp44744** [comment]
I fear it will initialize all the values in the form but I will try for sure

**Comment by swp44744** [comment]
I am trying to initialize my field as below, but its not working.  not sure if it is the correct way.
form declaration:
`export default reduxForm ({`
`  form: 'form_name',`
`  enableReinitialize: true,`
 ` validate`
`},null,null)(Class_Name);`

in separate component where I want to initialize field:
`initialize('form_name', {},['field_name']);`

Please suggest

**Comment by swp44744** [comment]
I am guessing this issue is related to below bug:
**[v6] Bug with data for fields which was generated dynamically by FieldArray in submit** #1685

**Comment by swp44744** [comment]
I was able to remove Field from "values" section by using action "change" as below:
Call change when you want to remove Field value:
`this.props.changeFieldValue('field_name','');`

in dispatchStateToProps:
`function mapDispatchToProps(dispatch) {`
 ` return {`
 `   changeFieldValue : function(field, value) {`
  `    dispatch(change('form_name', field, value))`
   ` }`
  `}`
`}`

Still working on removing from "fields(visited & touched)" section.

**Comment by erikras** [comment]
To remove the field from the page, you just stop rendering the `<Field>` element in your `render()` method. This will automatically call the `unregister` thing that @gustavohenke mentioned.

If you want to **set the value to `undefined`**, you can call `this.props.change('myNoLongerNeededField', undefined)`.

**Comment by oakley808** [comment]
I don't think unregistering the field clears out the `values` that the user might have entered in the field when present. And those orphan values will get submitted, which doesn't seem ideal.
(as of v6.2.1)

**Comment by swp44744** [comment]
@oakley808 use change action to set it to undefined and it will be gone from values section. Follow the solution mentioned by @erikras above. Either stop rendering it and it will unregister the field and if you want to get rid of it from values section then use 
`this.props.change('myNoLongerNeededField', undefined)`

**Comment by efdor** [comment]
@swp44744  @erikras,  I have tried:

`this.props.change('myNoLongerNeededField', undefined)`

and still see the old values upon submission of the form as @oakley808 described. I also tried 

`this.props.change('myNoLongerNeededField', "")`

before re-rendering the form (without the Fields) and the original values persist.

I'm using version 6.0.2

**Comment by gustavohenke** [comment]
Hi @efdor, can you try newer versions?
Redux Form became a lot more stable lately.

**Comment by crisys11** [comment]
this.props.change('myNoLongerNeededField', undefined) not works for me instead i use this.props.change('myNoLongerNeededField', null) this does not remove the field from  the store but remove the value.

**Comment by efdor** [comment]
setting the field to null is helpful for now until I get around to updating the library I'm using. Thanks!

**Comment by swp44744** [comment]
are you guys dispatching an change action, you might have to dispatch it from mapDispatchToProps as below:
`function mapDispatchToProps(dispatch) {`
 ` return {`
  `  changeFieldValue: function (field, value) {`
    `  dispatch(change(REDUX_FORM_NAME, field, value))`
`    }`
`  }`
`}`

Then use action as 
`this.props.changeFieldValue('myNoLongerNeededField', undefined)`

**Comment by dtipson** [comment]
You could also use reducer.plugin to detect the field unregistration event and remove the corresponding entry in the form's values state.  This seems to work for me.

**Comment by efdor** [comment]
We resolved this very simply by using the FieldArray component. It helped deal with our variable-length field sets without us having to add anything outside of the normal implementation.

**Comment by dtipson** [comment]
Unregistering a field also doesn't seem to update its entry in syncErrors, particularly for FormSection values.  

So when I try to display a `getFormSyncErrors` selector-based list of current errors to a user, that list will still include fields the user has already removed from the form, despite redux-form thinking that the form is now valid without them (i.e. for the sake of isValid, it ignores syncErrors for fields that aren't registered).  Unless I'm misunderstanding (or really messing things up somehow), I feel like syncErrors should stay in sync with registered fields directly, and isValid should coincide with an empty syncErrors store record.

**Comment by lloydh** [comment]
I just ran into this issue after trying to add/remove a field based on user input.

As efdor said, FieldArray is a good option but having been there it doesn't feel well suited when it's a single field being added/removed, and it also wraps the value in an unnecessary array.

I settled on the same `change(FORM, FIELD, null)` approach but this does leave null values to clean up. It would be nice to be able to 'destory' fields, including their values.

**Comment by borNfreee** [comment]
@erikras any news on this issue?
`FieldArray` does not suite the situation when field is not from collection.

**Comment by Sadnyster** [comment]
I'm having the same issue when trying to remove the field manually with `this.props.change('serverId', undefined)` but when I edit, the data from `selector(state, 'serverId')` still returned the old value which is wrong for me. I've tried set to `null` but still got the old value again and again. But if I set to a fixed value, it actually preserved the new value. Any idea guys?

**Comment by mmakarin** [comment]
change(field, undefined) is treated wrong in the latest version of redux-form.
```
if (initial === undefined && payload === '') {
      result = deleteInWithCleanUp(result, 'values.' + field);
    } else if (payload !== undefined) {
      result = setIn(result, 'values.' + field, payload);
    }
```
payload === undefined in this case => values remain in the state.
Please provide the way to remove the field explicitly (remove(form, field) action for example).

**Comment by jozsi** [comment]
I am also facing this issue. Changing a value to `undefined` leaves the old value in the store. Using v7.2.3.

**Comment by jozsi** [comment]
Dear @erikras - I am also facing this issue. Changing a value to undefined leaves the old value in the store. Using v7.2.3. Setting it to `null` is not a viable option for my use case.

**Comment by tmhammer** [comment]
+1 that this ought to be a feature.  It can make submit validation an unnecessary pain when there are fields with value `null` showing up when they don't need to be.

**Comment by rofrol** [comment]
`this.props.change('myNoLongerNeededField', undefined)` works with `redux-form 7.3.0`

**Comment by h0wXD** [comment]
We used a solution as follows iterate through the post values on submit, and delete any values set to null/undefined.
```
import { isNil } from 'ramda';

let newValues = {...values};
for(let key in newValues) {
    if (isNil(newValues[key])) {
        delete newValues[key];
    }
}

**Comment by Wgil** [comment]
`dispatch(autofill(ownProps.form, field, undefined))` worked for me

**Comment by alfimois** [comment]
> `dispatch(autofill(ownProps.form, field, undefined))` worked for me

Yeah, that's working! Thanks!

**Comment by kheengz** [comment]
`dispatch(autofill(ownProps.form, field, undefined))` worked for me. thanks @Wgil

**Comment by slava-tipper** [comment]
I am on v8
dispatch(autofill(ownProps.form, field, undefined)) worked for me also. Thanks

**Comment by AsuraKev** [comment]
where does autofill come from ?

**Comment by iamandrewluca** [comment]
@AsuraKev 
```
import { autofill } from 'redux-form'
```

**Comment by FlyingBurger** [comment]
I encounter an issue when I use autofill to remove a filed from redux.
My case is I want to remove the filed when the value of it becomes empty
```
     change('languages', this.state.languages);
     if (this.state.languages.length === 0) {
        dispatch(autofill(form, 'languages', undefined));
      }
```
when this function is triggered, the filed has been removed, and then the value is the same as the initial value, but the property 'dirty' is true.
I check the redux plugin, found 
![image](https://user-images.githubusercontent.com/53237722/75875102-f74e2900-5e0a-11ea-859d-60ae7c6af8c0.png)
![image](https://user-images.githubusercontent.com/53237722/75875122-02a15480-5e0b-11ea-9ba7-87fafefd34a6.png)
Is this field cause the form dirty? how can I get the form which is not dirty

**Comment by iamandrewluca** [comment]
I think you should remove the `<Field />` from the form when you want to remove it from redux, when a Field is removed from Form it will be unregistered.

**Comment by Zhenyria** [comment]
I use redux-form 8.3.7 and the problem seems is relevant.

I tried use following solutions:
```
dispatch(unregisterField(formName, fieldName, 'Field'));
dispatch(unregisterField(formName, fieldName));
dispatch(autofill(formName, fieldName, undefined));
dispatch(autofill(formName, fieldName, null));
dispatch(change(formName, fieldName, undefined));
dispatch(change(formName, fieldName, null));
```
but they didn't solve the problem
I have solved the problem with this soulution:
`dispatch(change(formName, fieldName, ''));`

It's really strange and it real delete the field

**Comment by rzippo** [comment]
The solutions reported here to be working seem to have been patched out with #3946.
If there is any way that is _meant_ to work, please document it
